File: src-tauri/src/cmd/path_tag.rs
use crate::{
    api::path_tag::PathTags, data::path_tag::entity, errors::AppError, state::ServiceAccess,
};

#[tauri::command]
pub fn create_path_tag(
    tag_id: &str,
    base_path: &str,
    source_id: &str,
    handle: tauri::AppHandle,
) -> Result<entity::PathTag, AppError> {
    handle.path_tag_ctrl(|ctrl| ctrl.create(tag_id, base_path, source_id))
}

#[tauri::command]
pub fn get_path_tag(
    tag_id: &str,
    base_path: &str,
    source_id: &str,
    handle: tauri::AppHandle,
) -> Result<entity::PathTag, AppError> {
    handle.path_tag_ctrl(|ctrl| ctrl.get_by_ids(tag_id, base_path, source_id))
}

#[tauri::command(async)]
pub fn list_path_tags(handle: tauri::AppHandle) -> Result<PathTags, AppError> {
    handle.path_tag_ctrl(|ctrl| ctrl.list()).map(PathTags::from)
}

#[tauri::command(async)]
pub fn list_path_tags_by_base_path(
    base_path: &str,
    source_id: &str,
    handle: tauri::AppHandle,
) -> Result<PathTags, AppError> {
    handle
        .path_tag_ctrl(|ctrl| ctrl.list_by_base_path(base_path, source_id))
        .map(PathTags::from)
}

#[tauri::command]
pub fn delete_path_tag(
    tag_id: &str,
    base_path: &str,
    source_id: &str,
    handle: tauri::AppHandle,
) -> Result<(), AppError> {
    handle.path_tag_ctrl(|ctrl| ctrl.delete(tag_id, base_path, source_id))
}

File: src/lib/api/path_tag.ts
import { invoke } from '@tauri-apps/api';
import type { TagAssignments } from './image_tag';

export interface PathTag {
	base_path: string;
	source_id: string;
	tag_id: string;
}

export interface PathTags {
	path_tags: PathTag[];
}

export const create = async (name: string, parentId: string): Promise<PathTag> => {
	const pathTag = (await invoke('create_path_tag', { name, parentId })) as PathTag;

	return pathTag;
};

export const list = async (): Promise<PathTags> => {
	const pathTags = (await invoke('list_path_tags', {})) as PathTags;

	return pathTags;
};

export const listByBasePath = async (basePath: string, sourceId: string): Promise<PathTags> => {
	const imageTags = (await invoke('list_path_tags_by_base_path', {
		basePath,
		sourceId
	})) as PathTags;

	return imageTags;
};

export const get = async (id: string): Promise<PathTag> => {
	const pathTag = (await invoke('get_path_tag', { id })) as PathTag;

	return pathTag;
};

export const remove = async (id: string): Promise<PathTag> => {
	const pathTag = (await invoke('delete_path_tag', { id })) as PathTag;

	return pathTag;
};

export const assign = async (
	basePaths: string[],
	sourceId: string,
	assignments: TagAssignments
) => {
	await invoke('assign_path_tags', {
		basePaths,
		sourceId,
		assignments
	});
};
